docs/tendermint-core/p2p/peer_manager.md

```diff
@@ -0,0 +1,439 @@
+# Peer manager
+
+The peer manager implements the connection policy for the node, based on the
+configuration provided by the operators, the current state of the connections
+reported by the [`Router`](./router.md), and the set of known candidate peers.
+
+This document uses *candidate peer* to refer to the information about a node in
+the network, namely its unique identifier and one or more network addresses.
+This information can be manually configured by the node operator (e.g., via
+`PersistentPeers` parameter) or can be obtained via the Peer-Exchange Protocol
+(PEX), which feeds the peer manager with discovered peers' information.
+
+## Connection policy
+
+The connection policy defines:
+
+1. When the node should establish new connections to peers, and
+1. The next peer to which the router should try to establish a connection.
+
+The first definition is made based the concept of [connection slots](#connection-slots).
+In short, the peer manager will try to fill every connection slot with a peer.
+If all the connection slots are full but there is the possibility to connect to
+a peer that is higher-[ranked](#peer-ranking) that one of the connecting peers,
+the peer manager may attempt to [upgrade a connection slot](#slot-upgrades).
+Details of these operations are provided in the following.
```

Comment:
This is a good point.

The main reason for having this is the concept of persistent peers. In the previous versions, they were not covered by the restrictions on the number of connections. With the peer manager, persistent peers are not treated as a different category of peers. Instead, they have their score set in a definitive way to the maximum value. As a result, whenever a persistent peer connects to us, or we can connect to a persistent peer, it will take the connection slot of a "regular" peer.